Sources/Test/SimpleTesting/Aggregates/HoppingWindowTopKAggregate.cs

```diff
@@ -0,0 +1,204 @@
+        public void TestHoppingWindowTopKAggregateAllIncreasing()
+        {
+            GenerateDataAndTestInput(
+                numValues: 1000,
+                valueGenerator: v => v + random.Next(1, 11),
+                distanceGenerator: () => HopSize);
+        }
+
+        [TestMethod, TestCategory("Gated")]
+        public void TestHoppingWindowTopKAggregateAllDecreasing()
+        {
+            GenerateDataAndTestInput(
+                numValues: 1000,
+                valueGenerator: v => (v == 0) ? 10000 : v - random.Next(1, 11),
+                distanceGenerator: () => HopSize);
+        }
+
+        [TestMethod, TestCategory("Gated")]
+        public void TestHoppingWindowTopKAggregateCyclingValues()
+        {
+            // Test values 1 -> 2 -> 3 -> 1, and run for 1/2, 1, 2, 4, 8 intervals of hops
+            for (long distance = HopSize / 2; distance <= HopSize * 8; distance *= 2)
+            {
+                GenerateDataAndTestInput(
+                    numValues: 10,
+                    valueGenerator: v => (v % 3) + 1,
+                    distanceGenerator: () => distance);
+            }
+        }
+
+        private void GenerateDataAndTestInput(
+            int numValues,
+            Func<int, int> valueGenerator,
+            Func<long> distanceGenerator,
+            long windowSize = WindowSize)
+        {
+            var input = new List<StreamEvent<int>>();
+            long maxStartTime = 0;
+
+            long startTime = 0;
+            int value = 100;
+            for (int i = 0; i < numValues; i++)
+            {
+                startTime += distanceGenerator();
+                value = valueGenerator(value);
+                input.Add(StreamEvent.CreateStart(startTime, value));
+                maxStartTime = Math.Max(maxStartTime, startTime);
+            }
+            input.Add(EndEvent);
+
+            TestHoppingWindowTopKAggregateInternal(input, maxStartTime, windowSize);
+        }
+
+        private void TestHoppingWindowTopKAggregateInternal(IEnumerable<StreamEvent<int>> streamEvents, long maxStartTime, long windowSize)
+        {
+            var output = streamEvents.ToStreamable().HoppingWindowLifetime(windowSize, HopSize).TopK(K);
+
+            var correct = new List<StreamEvent<List<int>>>();
+
+            maxStartTime += windowSize;
+            for (long startTime = 0; startTime < maxStartTime; startTime += HopSize)
+            {
+                var eventsInWindow = streamEvents.Where(x => x.StartTime > (startTime - windowSize) && x.StartTime <= startTime);
+
+                if (!eventsInWindow.Any())
+                    continue;
+
+                var dataInWindow = eventsInWindow.Select(se => se.Payload).ToList();
+                dataInWindow.Sort((x, y) => y.CompareTo(x));
+
+                var last = dataInWindow[Math.Min(K, dataInWindow.Count) - 1];
+                var topKInWindow = dataInWindow.TakeWhile(e => e >= last).ToList();
```

Comment:
Does this handle duplicates in the TopK ranks?

Reply:
Yes, by doing take-last and then TakeWhile(e => e== last)